mypy/newsemanal/semanal_main.py

```diff
@@ -168,10 +168,12 @@ def process_top_levels(graph: 'Graph', scc: List[str], patches: Patches) -> None
         worklist += list(reversed(core_modules)) * CORE_WARMUP
     final_iteration = False
     iteration = 0
+    analyzer = state.manager.new_semantic_analyzer
+    analyzer.deferral_debug_context.clear()
+
     while worklist:
         iteration += 1
         if iteration > MAX_ITERATIONS:
-            analyzer = state.manager.new_semantic_analyzer
             # Just pick some module inside the current SCC for error context.
             assert state.tree is not None
             with analyzer.file_context(state.tree, state.options):
```

Comment:
Hang reporting doesn't necessarily report the correct information here, since we only report the deferral context for some module in the SCC, which might not even be participating in the final iterations.

Reply:
But we should set _some_ context. The trace printed above makes it clear (assuming the import cycle bug is fixed, I am doing this now).

Reply:
I was mistaken about this. I was confused by the code accessing the semantic analyzer through the state object. It might be worth it to refactor it -- I can look at it.